File: tests/c/test_vfs_exchange.c
/*
 * test_vfs_exchange.c — the phase-107 C6 atomic two-name exchange, against the
 * REAL namespace_ops.o + beneath.o (renameat2(RENAME_EXCHANGE) through the
 * confined rootfd is the point: the security cases only mean something with
 * the kernel confinement actually live).
 *
 * No protocol verb reaches brix_vfs_exchange yet (the OCI tag flip is
 * phase-108 C10), so this unit is the ONLY behavior coverage for the verb.
 * The contract it pins (namespace_ops.c, §3.5 of the phase doc):
 *   - both names swap in ONE kernel op — witnessed by inode identity, not
 *     content copy (a two-rename emulation would mint new inodes / windows);
 *   - ENOENT unless BOTH names exist (matching renameat2), survivor untouched;
 *   - either name outside the export root refuses EXDEV → BRIX_NS_DENIED,
 *     never a confinement escape — including the "/root"-vs-"/rootevil"
 *     prefix-boundary trap in brix_beneath_strip_root;
 *   - the primitive is type-agnostic (file <-> directory swaps work).
 *
 * Cross-TU stubs: the impersonation broker (inactive), and namespace_ops.o's
 * delete/mkdir-path externs (unreachable from exchange — abort bodies prove
 * it). ngx_cycle/ngx_log_error_core come from the shared ngx_link_stubs.c.
 */
#include <ngx_config.h>
#include <ngx_core.h>

#include "core/compat/namespace_ops.h"
#include "fs/path/beneath.h"

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>

/* ---- cross-TU stubs ----------------------------------------------------- */

int brix_imp_client_active(void) { return 0; }
int brix_imp_open(const char *p, int f, mode_t m) { (void) p; (void) f; (void) m; abort(); }
int brix_imp_stat(const char *p, struct stat *st, int nf) { (void) p; (void) st; (void) nf; abort(); }
int brix_imp_mkdir(const char *p, mode_t m) { (void) p; (void) m; abort(); }
int brix_imp_unlink(const char *p, int d) { (void) p; (void) d; abort(); }
int brix_imp_rename(const char *s, const char *d) { (void) s; (void) d; abort(); }
int brix_imp_rename_noreplace(const char *s, const char *d) { (void) s; (void) d; abort(); }
int brix_imp_link(const char *s, const char *d) { (void) s; (void) d; abort(); }

/* namespace_ops.o externs from verbs this unit never drives — abort bodies
 * double as the proof that exchange stays off the delete/mkdir planes. */
int
brix_mkdir_recursive_beneath(ngx_log_t *log, int rootfd,
    const char *root_canon, const char *resolved, mode_t mode,
    ngx_array_t *rules)
{
    (void) log; (void) rootfd; (void) root_canon; (void) resolved;
    (void) mode; (void) rules;
    abort();
}

ngx_int_t
brix_fs_dir_is_empty(const char *path, ngx_flag_t *is_empty)
{
    (void) path; (void) is_empty;
    abort();
}

ngx_int_t
brix_fs_remove_tree_confined(ngx_log_t *log, const char *root_canon,
    const char *path)
{
    (void) log; (void) root_canon; (void) path;
    abort();
}

/* ---- helpers ------------------------------------------------------------ */

static char       g_root[128];   /* short so "%s/%s" composes warning-free */
static ngx_log_t  g_log;

static void
put_file(const char *path, const char *content)
{
    FILE *f = fopen(path, "w");

    assert(f != NULL);
    assert(fputs(content, f) >= 0);
    assert(fclose(f) == 0);
}

static void
expect_file(const char *path, const char *content)
{
    char  buf[64];
    FILE *f = fopen(path, "r");
    size_t n;

    assert(f != NULL);
    n = fread(buf, 1, sizeof(buf) - 1, f);
    buf[n] = '\0';
    assert(strcmp(buf, content) == 0);
    assert(fclose(f) == 0);
}

static ino_t
ino_of(const char *path)
{
    struct stat sb;

    assert(lstat(path, &sb) == 0);
    return sb.st_ino;
}

int
main(void)
{
    char              a[PATH_MAX], b[PATH_MAX], buf[PATH_MAX];
    ino_t             ino_a, ino_b;
    brix_ns_result_t  res;

    g_log.log_level = NGX_LOG_DEBUG;
    snprintf(g_root, sizeof(g_root), "/tmp/xchg-%d", (int) getpid());
    assert(mkdir(g_root, 0755) == 0);
    snprintf(buf, sizeof(buf), "%s/d", g_root);
    assert(mkdir(buf, 0755) == 0);

    /* 1. SUCCESS: file <-> file across directories, one kernel op. The inode
     * swap is the atomicity witness: a two-rename emulation (banned by §3.5)
     * would expose an ENOENT window and could not preserve both inodes. */
    snprintf(a, sizeof(a), "%s/a.txt", g_root);
    snprintf(b, sizeof(b), "%s/d/b.txt", g_root);
    put_file(a, "alpha");
    put_file(b, "bravo!");
    ino_a = ino_of(a);
    ino_b = ino_of(b);
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_OK);
    assert(res.sys_errno == 0);
    expect_file(a, "bravo!");
    expect_file(b, "alpha");
    assert(ino_of(a) == ino_b);
    assert(ino_of(b) == ino_a);

    /* 2. SUCCESS: the primitive is type-agnostic — a regular file and a
     * populated directory swap names; the directory's child rides along. */
    snprintf(a, sizeof(a), "%s/plain", g_root);
    snprintf(b, sizeof(b), "%s/subdir", g_root);
    put_file(a, "plain-bytes");
    assert(mkdir(b, 0755) == 0);
    snprintf(buf, sizeof(buf), "%s/subdir/child", g_root);
    put_file(buf, "child-bytes");
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_OK);
    snprintf(buf, sizeof(buf), "%s/plain/child", g_root);
    expect_file(buf, "child-bytes");        /* dir now answers to "plain" */
    expect_file(b, "plain-bytes");          /* file now answers to "subdir" */

    /* 3. ERROR: one name missing → ENOENT/NOT_FOUND (renameat2 semantics:
     * BOTH must exist), and the survivor is untouched — same inode, same
     * bytes. Exchange must never half-create the missing side. */
    snprintf(a, sizeof(a), "%s/a.txt", g_root);
    snprintf(b, sizeof(b), "%s/never-made", g_root);
    ino_a = ino_of(a);
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_NOT_FOUND);
    assert(res.sys_errno == ENOENT);
    assert(ino_of(a) == ino_a);
    expect_file(a, "bravo!");
    assert(access(b, F_OK) == -1);

    /* 4. ERROR: both names missing is still plain ENOENT, not a crash or a
     * spurious create. */
    snprintf(a, sizeof(a), "%s/ghost1", g_root);
    snprintf(b, sizeof(b), "%s/ghost2", g_root);
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_NOT_FOUND);
    assert(res.sys_errno == ENOENT);

    /* 5. SECURITY-NEGATIVE: first name outside the export root → EXDEV /
     * DENIED before any syscall touches it; the outside file survives byte
     * for byte. (Cross-EXPORT exchange is refused at the VFS layer the same
     * way — one rootfd confines both ends here.) */
    snprintf(buf, sizeof(buf), "/tmp/xchg-outside-%d", (int) getpid());
    assert(mkdir(buf, 0755) == 0);
    snprintf(a, sizeof(a), "%s/loot", buf);
    put_file(a, "outside");
    snprintf(b, sizeof(b), "%s/a.txt", g_root);
    ino_b = ino_of(b);
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_DENIED);
    assert(res.sys_errno == EXDEV);
    expect_file(a, "outside");
    assert(ino_of(b) == ino_b);

    /* 6. SECURITY-NEGATIVE: second name outside the root refuses identically
     * (the ns_rel arm rather than the strip_root arm). */
    res = brix_ns_exchange(&g_log, g_root, b, a);
    assert(res.status == BRIX_NS_DENIED);
    assert(res.sys_errno == EXDEV);
    expect_file(a, "outside");

    /* 7. SECURITY-NEGATIVE: the prefix-boundary trap — "/tmp/xchg-N-evil"
     * shares the root's string prefix but is NOT under it; strip_root must
     * reject at the '/' boundary, not on strncmp alone. */
    snprintf(buf, sizeof(buf), "%s-evil", g_root);
    assert(mkdir(buf, 0755) == 0);
    snprintf(a, sizeof(a), "%s-evil/f", g_root);
    put_file(a, "evil");
    res = brix_ns_exchange(&g_log, g_root, a, b);
    assert(res.status == BRIX_NS_DENIED);
    assert(res.sys_errno == EXDEV);
    expect_file(a, "evil");

    printf("vfs_exchange: 7 cases ok\n");
    return 0;
}

File: tests/test_vfs_exchange.py
"""test_vfs_exchange.py — phase-107 C6: the atomic two-name exchange contract.

Behavior coverage lives in the C object unit (tests/c/test_vfs_exchange.c —
real renameat2(RENAME_EXCHANGE) through the real confined rootfd: inode-swap
atomicity, both-must-exist ENOENT, EXDEV confinement including the
prefix-boundary trap).  No protocol verb reaches brix_vfs_exchange yet (the
OCI tag flip is phase-108), so what this file pins is the SOURCE contract the
verb must keep while it waits for its first caller:

  - the typed policy gate runs FIRST — before the cross-export comparison —
    so a read-only export answers EROFS without leaking whether either name
    exists (the same existence-probe rule the conditional-publish rows carry);
  - cross-export exchange is EXDEV at the VFS layer (the driver arm's
    export-relative keying would otherwise pass a foreign path through);
  - a backend without the primitive refuses ENOTSUP and is NEVER emulated
    with two renames (§3.5) — in the dispatch helper AND in both engine arms;
  - the C3 durable-publish barrier and the cache eviction cover BOTH names
    (both directory entries changed);
  - both decorators relay the slot pair (the R-wave parity lesson: a slot the
    cache does not relay turns a working leaf primitive into ENOTSUP).

Reference: docs/refactor/phase-107-vfs-mutation-surface-completion.md §3.5/C6.

Run:
  PYTHONPATH=tests pytest tests/test_vfs_exchange.py -v
"""

import pathlib
import re

REPO = pathlib.Path(__file__).resolve().parents[1]


def _src(rel: str) -> str:
    return (REPO / rel).read_text(encoding="utf-8")


def _fn(text: str, name: str) -> str:
    """The body of a top-level C function: from its definition line to the
    first closing brace at column 0."""
    m = re.search(rf"^{re.escape(name)}\(.*?^\}}", text, re.S | re.M)
    assert m, f"function {name} not found — re-anchor this pin"
    return m.group(0)


# --------------------------------------------------------------------------
# The VFS verb: gate order, EXDEV, no emulation
# --------------------------------------------------------------------------

def test_policy_gate_runs_before_any_name_comparison():
    """EROFS before EXDEV/EINVAL: a read-only export must refuse the mutation
    without evaluating either name, or the refusal ordering becomes an
    existence/confinement probe.  The gate lives in the shared entry helper
    (brix_vfs_two_name_entry); the pin follows it there AND pins the helper
    call above the EXDEV check in brix_vfs_exchange itself."""
    text = _src("src/fs/vfs/vfs_rename.c")
    entry = _fn(text, "brix_vfs_two_name_entry")
    assert "brix_vfs_require_confined_mutation" in entry, (
        "the typed policy gate left the shared entry helper")
    assert "BRIX_VFS_MUTATE_RENAME" in entry, (
        "exchange is write-gated as a rename-class mutation")
    body = _fn(text, "brix_vfs_exchange")
    gate = body.index("brix_vfs_two_name_entry(")
    xdev = body.index("brix_beneath_strip_root")
    assert gate < xdev, "the entry gate moved below the EXDEV check"


def test_cross_export_exchange_is_exdev_before_driver_dispatch():
    body = _fn(_src("src/fs/vfs/vfs_rename.c"), "brix_vfs_exchange")
    m = re.search(r"brix_beneath_strip_root\(.*?errno = EXDEV;", body, re.S)
    assert m, "the cross-export EXDEV refusal left brix_vfs_exchange"
    assert body.index("errno = EXDEV") < body.index(
        "brix_vfs_exchange_driver(ctx"), (
        "EXDEV must be decided before the driver arm sees a foreign key")


def test_exchange_is_never_emulated_with_two_renames():
    """§3.5: a caller that asked for an atomic swap would rather have ENOTSUP
    than a window in which neither name resolves.  Neither engine arm may
    fall back to rename."""
    ns = _fn(_src("src/core/compat/namespace_ops.c"), "brix_ns_exchange")
    assert "brix_exchange_beneath" in ns
    assert "brix_rename_beneath" not in ns, (
        "the POSIX arm grew a rename fallback — §3.5 forbids emulation")

    dispatch = _src("src/fs/backend/sd_cred_forward.h")
    m = re.search(
        r"if \(inst->driver->exchange == NULL\) \{.*?errno = ENOTSUP;",
        dispatch, re.S)
    assert m, "the NULL-slot ENOTSUP refusal left brix_sd_exchange_maybe_cred"

    drv = _fn(_src("src/fs/vfs/vfs_rename.c"), "brix_vfs_exchange_driver")
    assert "brix_sd_exchange_maybe_cred" in drv
    assert "->rename(" not in drv and "brix_sd_rename" not in drv, (
        "the driver arm grew a rename fallback — §3.5 forbids emulation")


# --------------------------------------------------------------------------
# Both names: barrier + eviction
# --------------------------------------------------------------------------

def test_c3_barrier_covers_both_names_in_both_arms():
    text = _src("src/fs/vfs/vfs_rename.c")
    posix_arm = _fn(text, "brix_vfs_exchange")
    assert len(re.findall(r"brix_publish_dirsync\(", posix_arm)) == 2, (
        "the POSIX arm must flush BOTH parents — both entries changed")
    driver_arm = _fn(text, "brix_vfs_exchange_driver")
    assert len(re.findall(r"sync_publish\(leaf, [ab]_key\)", driver_arm)) == 2, (
        "the driver arm must barrier BOTH keys")


def test_driver_arm_evicts_both_keys():
    driver_arm = _fn(_src("src/fs/vfs/vfs_rename.c"),
                     "brix_vfs_exchange_driver")
    assert re.search(
        r"brix_sd_cache_evict\(ctx->sd, a_key\)\s*\+"
        r"\s*brix_sd_cache_evict\(ctx->sd, b_key\)", driver_arm), (
        "both keys map to new content — both cache entries must go")


# --------------------------------------------------------------------------
# Decorator parity (the R-wave lesson)
# --------------------------------------------------------------------------

def test_both_decorators_relay_the_exchange_slot_pair():
    for rel in ("src/fs/backend/cache/sd_cache.c",
                "src/fs/backend/stage/sd_stage.c"):
        text = _src(rel)
        assert re.search(r"\.exchange\s*=", text), f"{rel}: .exchange relay gone"
        assert re.search(r"\.exchange_cred\s*=", text), (
            f"{rel}: .exchange_cred relay gone — the cred twin is the "
            f"security-relevant half")


def test_posix_leaf_implements_the_primitive():
    text = _src("src/fs/backend/posix/sd_posix.c")
    assert re.search(r"\.exchange\s*=\s*sd_posix_exchange", text)
